core/src/main/kotlin/org/evomaster/core/search/gene/FlexibleObjectGene.kt

```diff
@@ -0,0 +1,171 @@
+package org.evomaster.core.search.gene
+
+import org.evomaster.core.logging.LoggingUtil
+import org.evomaster.core.output.OutputFormat
+import org.evomaster.core.search.gene.collection.PairGene
+import org.evomaster.core.search.gene.optional.FlexibleGene
+import org.evomaster.core.search.gene.root.CompositeGene
+import org.evomaster.core.search.gene.string.StringGene
+import org.evomaster.core.search.gene.utils.GeneUtils
+import org.evomaster.core.search.gene.utils.GeneUtils.isInactiveOptionalGene
+import org.evomaster.core.search.service.AdaptiveParameterControl
+import org.evomaster.core.search.service.Randomness
+import org.evomaster.core.search.service.mutator.MutationWeightControl
+import org.evomaster.core.search.service.mutator.genemutation.AdditionalGeneMutationInfo
+import org.evomaster.core.search.service.mutator.genemutation.SubsetGeneMutationSelectionStrategy
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+class FlexibleObjectGene<T>(
```

Comment:
add some documentation for this gene.

also, i guess there is quite a bit of code that would be similar to `ObjectGene`.

Wondering if having a new gene that internally contains an `ObjectGene` (for the fixed fields) and a `MapGene` (for the additional properties) would be cleaner. Or simply adding a `MapGene` directly inside `ObjectGene` as optional.

Wondering about all places in the code in which we have checks like `is ObjectGene` which will not work for this new gene

Reply:
the reason why I did not extend it from `ObjectGene` is that `ObjectGene` is _fixed_ composite gene.
We might need to extract `ObjectGene` as `interface`, then further distinguish `FixedObjextGene` and `FlexiableObjectGene`. 
However, for `FixedObjextGene` and `ObjectGene`, most of methods (eg, methods related to copy, and mutation) are different.

`FlexiableObjectGene` was not defined with `ObjectGene` and `MapGene` because of its structure of children. currently, the `children` is a list of fields  (fixed fields plus additional ones). the structure of children might relate to mutation. 

another solution (which might not be clean) is that we extend `FixedObjextGene` from `ObjectGene`, and make it have fixed two fields, i.e., _fixedFields:`ObjectGene`_ and _additionalFields:`MapGene`_.  thus, we could identify `FixedObjextGene` as `FixedCompositegene`. but in this case, we make the structure of the gene a bit complex.